pkg/tags/tag.go

```diff
@@ -172,11 +172,15 @@ func (t *Tag) Done(s State) bool {
 	return err == nil && n == total
 }
 
-// DoneSplit sets total count to SPLIT count and sets the associated swarm hash for this tag
-// is meant to be called when splitter finishes for input streams of unknown size
+// DoneSplit adds the total with the split count and updated the total.
+// This is useful when you use the same tag for a file and its manifest upload
+// Also the Address is updated with the last called address, so the assumption
+// is that the manifest creation will be called last and this will have the root
+// hash of the manifest
 func (t *Tag) DoneSplit(address swarm.Address) int64 {
-	total := atomic.LoadInt64(&t.Split)
-	atomic.StoreInt64(&t.Total, total)
+	split := atomic.LoadInt64(&t.Split)
+	total := atomic.LoadInt64(&t.Total)
+	atomic.StoreInt64(&t.Total, total+split)
```

Comment:
These three atomic functions done separately are not atomic. There is a possibility of data race, as nothing is guarding Total in between lines 182 and 183. Total got on line 182 may be changed before new value is stored on line 183 by some other goroutine resulting an incorrect value. Mutex should be used.

Reply:
Your change is fixing the issue with minimal set of changes, but I find at least questionable whether both atomic and mutex should be used at the same time. If there would be any other place where Total would need to be set, both mutex lock should be held and atomic function used. There is even a strange code in UnmarshalBinary that sets Total with assignment, which should be set with an atomic function, but this is out of the scope of this pr.

Reply:
how else do you propose to fix this in this place? Other places a okay as they dont take it from splitCount.

The actual way to fix this would be to increment the total counter in a place somewhere between the API and the chunker. In the previous code it was the hasher...now we have to think it when we do pyramid chunker or something.

Reply:
This is a completely local issue of protecting the memory. I would not use atomic for Total field, just mutex as it is required here and makes it obvious how it is protected. I know that this involves a few more lines to be changed, but it makes clear how Total value should be handled. This is the danger of overusing the atomic package as you can create a false sense of data race free code.

Reply:
i am not sure if we should not use atomic for Total.... when other use cases comes in we might have to increase them atomically.